ocaml/src/fragment_cache.ml

```diff
@@ -22,20 +22,31 @@ open Lwt_bytes2
 module KV = Rocks_key_value_store
 
 
-class type cache = object
-    method add : int32 -> string -> Lwt_bytes.t -> unit Lwt.t
-    method lookup : int32 -> string -> Lwt_bytes.t option Lwt.t
-    method lookup2 : int32 -> string -> (int * int * Lwt_bytes.t * int) list -> bool Lwt.t
+class virtual cache ~cache_on_read ~cache_on_write = object(self)
 
-    method drop  : int32 -> global : bool -> unit Lwt.t
-    method close : unit -> unit Lwt.t
+    method virtual _add : int32 -> string -> Bigstring_slice.t -> unit Lwt.t
+    method add bid key blob (moment : [ `Read | `Write ]) =
+      let cache = function
+        | true -> self # _add bid key blob
+        | false -> Lwt.return_unit
+      in
+      match moment with
+      | `Read -> cache cache_on_read
+      | `Write -> cache cache_on_write
```

Comment:
I think I wrote it like this to ensure exhaustive matching on 'moment'.

Regarding the other part of your comment ... there is indeed something a bit odd.
I think it's because the fragment cache shouldn't care where the add comes from. So pushed the decision wether or not to cache out to the caller in a new commit.